### PR TITLE
Invalidate resource references when the referenced resource is moved

### PR DESCRIPTION
**Pull request by SupunS**
## Description

This is the draft implementation of the FLIP: https://github.com/onflow/flow/pull/1043

The general idea is that, a resource reference is invalidated, if the pointed resource is being moved.

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/value.go [hunk 1]
```diff
@@ -17556,6 +17577,7 @@ func (v *EphemeralReferenceValue) RemoveKey(
 	self := *referencedValue
 
 	interpreter.checkReferencedResourceNotDestroyed(self, getLocationRange)
+	v.checkReferenceResourceNotMoved(self, getLocationRange)
```
Comment (dsainati1) [inline runtime/interpreter/value.go:17580]: These seem to occur together all the time; can we refactor this to be one function call that does both checks?

Comment (SupunS) [reply runtime/interpreter/value.go:17580]: Yeah, should be possible. I haven't migrated the existing tests, so wasn't sure what other places need updates, etc. So kept it separated for now. But I guess we can merge the two where ever possible.

File: runtime/interpreter/errors.go [hunk 2]
```diff
@@ -865,3 +865,18 @@ func (InvalidHexLengthError) IsUserError() {}
...
+type MovedResourceReferenceError struct {
+	LocationRange
+}
+
+var _ errors.UserError = MovedResourceReferenceError{}
+
+func (MovedResourceReferenceError) IsUserError() {}
+
+func (e MovedResourceReferenceError) Error() string {
+	return "referring resource is moved"
```
[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/interpreter/errors.go:881]: That's a good idea 👍 I'll see if we can point to the location where it was moved

[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/interpreter/errors.go]: I had a look into this, and it seems very difficult to keep track of the 'move' operation that causes the invalidation.
e.g: A single resource can have multiple references and multiple move operations. Also, the move operations are not aware of the existing references (we don't maintain inward references to the resource). So if we are to add this, need to maintain such mapping separately, which can be costly. That kind of bookkeeping is a bit too heavy for an error message improvement.

[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/interpreter/errors.go]: improved the error message slightly https://github.com/onflow/cadence/pull/1961/commits/a6250a67cbcd9be89e5a7de0d21e9c2293aaa547

File: runtime/tests/interpreter/reference_test.go [hunk 3]
```diff
@@ -472,23 +472,17 @@ func TestInterpretResourceReferenceAfterMove(t *testing.T) {
...
 		)
 
-		arrayRef := &interpreter.EphemeralReferenceValue{
-			Authorized: false,
-			Value:      array,
-			BorrowedType: &sema.VariableSizedType{
+		arrayRef := interpreter.NewUnmeteredEphemeralReferenceValue(
+			false,
+			array,
+			&sema.VariableSizedType{
```
Comment (dreamsmasher) [inline runtime/tests/interpreter/reference_test.go:478]: maybe keep the labels here, although it's just a test case
```suggestion
		arrayRef := interpreter.NewUnmeteredEphemeralReferenceValue(
			Authorized:   false,
			Value:        array,
			BorrowedType: &sema.VariableSizedType{
```

**Review by dreamsmasher — APPROVED** [review]

File: runtime/tests/interpreter/reference_test.go (hunk 3 above)
Comment (SupunS) [reply runtime/tests/interpreter/reference_test.go:478]: This was changed from the struct-literal to a function call (a constructor). So the labels are gone now.

**Comment by SupunS** [comment]
The behaviour of the implementation was slightly modified to invalidate the resource immediately upon a move, as described in https://github.com/onflow/flow/pull/1043#issuecomment-1247394376. I'll open a separate PR with this modified behaviour. Hence closing this one.

**Comment by SupunS** [comment]
New PR: https://github.com/onflow/cadence/pull/1999

**Pull request closed without merging**